### PR TITLE
Add default retry for load config via url

### LINKED ISSUE #7338: Agent won't start if remote HTTP config endpoint is down
## Feature Request
A useful feature of Telegraf is to pull config from a remote HTTP server, thus centralising agent configuration.  However, if the HTTP endpoint is unavailable when the agent is started, the agent throws an error and dies.  

### Proposal:
Have the agent retry the endpoint periodically if unavailable, and when online, pull the config and start as normal.

### Current behavior:
Agent will not start.

### Desired behavior:
Agent re-attempts to get the configuration, ideally driven by startup switches to control retry-interval and max retry-attempts

### Use case: 
In a large environment, it is entirely possible that the HTTP endpoint could go offline (e.g. planned or unplanned outage).  Any agents that get restarted during this interval (e.g. reboot, agent installation) would simply die, meaning some recovery process is needed to bring them back online.  
Having this retry ability in the agent would provide a more reliable solution and self-recovery.

### PR DESCRIPTION
**Pull request by helenosheaa**
Closes #7338

Related #7349, #7662 

This is a temporary fix which we intend to improve on:

- Retries three times at 10s intervals when receiving an error on loading config from a url incase of the endpoint being down.

- Log attempts and the specific failure response codes.

- Does not use env variables or add new flags 

- Tests added to demonstrate that on the final failure it will return the error.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -921,17 +922,32 @@ func fetchConfig(u *url.URL) ([]byte, error) {
...
 	req.Header.Add("Accept", "application/toml")
 	req.Header.Set("User-Agent", internal.ProductToken())
-	resp, err := http.DefaultClient.Do(req)
-	if err != nil {
-		return nil, err
-	}
 
-	if resp.StatusCode != http.StatusOK {
-		return nil, fmt.Errorf("failed to retrieve remote config: %s", resp.Status)
+	retries := 3
```
Comment (Jagularr) [inline config/config.go:926]: nitpick but I guess this (retries) could be added to the constants at the top too, for consistency

**Review by Jagularr — APPROVED** [review]

File: config/config.go (hunk 1 above)
Comment (helenosheaa) [reply config/config.go:926]: Yeah I get you. I only added the interval var up top just so I could access and change it for the test so it wouldn't take 30s to run

File: config/config.go [hunk 2]
```diff
@@ -921,17 +922,32 @@ func fetchConfig(u *url.URL) ([]byte, error) {
...
+	for i := 1; i <= retries; i++ {
```
Comment (ssoroka) [inline config/config.go:927]: if retries is 0, this won't attempt at all. If we later move to user-configurable retries, that won't quite make sense

File: config/config.go [hunk 3]
```diff
@@ -921,17 +922,32 @@ func fetchConfig(u *url.URL) ([]byte, error) {
...
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
```
Comment (ssoroka) [inline config/config.go:929]: non-200 responses don't return errors, so we might want to not retry on err. err could indicate something more fundamental is wrong. Check the docs for what conditions Do returns errors.

File: config/config.go (hunk 2 above)
Comment (helenosheaa) [reply config/config.go:927]: Right changed i to 0

File: config/config.go (hunk 3 above)
Comment (helenosheaa) [reply config/config.go:929]: So it only returns on error either for client policy or failure to speak HTTP. So the use case was for when someones network is temporarily down I'm thinking the retries on http status code covers that so will just break if there is an err

**Review by ssoroka — APPROVED** [review]

**Pull request merged**